From a425973ac4e65dde4486e66b000ab52cd6b4cb31 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 27 Mar 2012 16:07:35 +0800
Subject: [PATCH 1/4] netdev/dpa: add a check before deleting the pcd of a
 port

Add a check to verify if the pcd is enabled on this port
before deleting it. This will fix an oops like the following
because the h_FmPcd is null when pcd is not enabled yet on
that port.

Unable to handle kernel paging request for data at address 0x0000005e
Faulting instruction address: 0xc0471530
Oops: Kernel access of bad area, sig: 11 [#1]
PREEMPT SMP NR_CPUS=4 LTT NESTING LEVEL : 0
P2041 RDB
last sysfs file: /sys/devices/virtual/block/md0/dev
Modules linked in: x_tables ip_tables ipv6 sctp binfmt_misc [last
unloaded: scsi_wait_scan]
NIP: c0471530 LR: c048ca1c CTR: c04a9240
REGS: eb3a7d10 TRAP: 0300   Not tainted  (2.6.34.10-WR4.3.0.0_standard)
MSR: 00029002 <EE,ME,CE>  CR: 24000422  XER: 20000000
DEAR: 0000005e, ESR: 00000000
TASK = eb2f1500[700] 'fmc' THREAD: eb3a6000 CPU: 0
GPR00: c048ca1c eb3a7dc0 eb2f1500 00000000 00000000 00000000 00000000
81010100
GPR08: 00000003 00000000 c011b204 00000000 00000000 1012144c bfe98e00
bfe98fdc
GPR16: bfe9906c 100d7e98 bfe990a0 100dc150 bfe990d4 00000000 bfe98d78
00000000
GPR24: 1015d640 00000000 bfe98d7c 2000e147 0000000c 00000050 00000000
00000000
NIP [c0471530] FmPcdDecNetEnvOwners+0x1c/0x84
LR [c048ca1c] FM_PORT_DeletePCD+0x190/0x36c
Call Trace:
[eb3a7dc0] [eb8a9880] 0xeb8a9880 (unreliable)
[eb3a7de0] [c048ca1c] FM_PORT_DeletePCD+0x190/0x36c
[eb3a7e30] [c04aae10] LnxwrpFmPortIOCTL+0x4fc/0xf40
[eb3a7e60] [c04a9344] fm_ioctl+0x104/0x140
[eb3a7e80] [c011a8b4] vfs_ioctl+0x3c/0xec
[eb3a7ea0] [c011ab2c] do_vfs_ioctl+0x80/0x758
[eb3a7f10] [c011b298] sys_ioctl+0x94/0x108
[eb3a7f40] [c00105dc] ret_from_syscall+0x0/0x4

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../dpa/NetCommSw/Peripherals/FM/Port/fm_port.c    |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/dpa/NetCommSw/Peripherals/FM/Port/fm_port.c b/drivers/net/dpa/NetCommSw/Peripherals/FM/Port/fm_port.c
index 7a8434c..2382bf5 100644
--- a/drivers/net/dpa/NetCommSw/Peripherals/FM/Port/fm_port.c
+++ b/drivers/net/dpa/NetCommSw/Peripherals/FM/Port/fm_port.c
@@ -4549,6 +4549,9 @@ t_Error FM_PORT_DeletePCD(t_Handle h_FmPort)
     if (p_FmPort->imEn)
         RETURN_ERROR(MAJOR, E_INVALID_OPERATION, ("available for non-independant mode ports only"));
 
+    if (!p_FmPort->pcdEngines)
+	RETURN_ERROR(MAJOR, E_INVALID_OPERATION, ("called for non PCD port"));
+
     if(p_FmPort->pcdEngines & FM_PCD_KG)
     {
         /* unbind all schemes */
-- 
1.7.9.7

